packages/core/test/VDocument.test.ts

```diff
@@ -1,9 +1,169 @@
     describe('VDocument', () => {
+        describe('withCustomRange', () => {
+            it('should work with a provided VRange ', () => {
+                testEditor({
+                    contentBefore: 'ab[]',
+                    stepFunction: async (editor: JWEditor) => {
+                        await editor.vDocument.withCustomRange(editor.vDocument.range, range => {
+                            editor.vDocument.insertText('c', range);
+                            editor.renderer.render(editor.vDocument, editor.editable);
+                        });
+                    },
+                    contentAfter: 'abc[]',
+                });
+            });
+            it('should take the current range as default', () => {
+                testEditor({
+                    contentBefore: 'ab[]',
+                    stepFunction: async (editor: JWEditor) => {
+                        await editor.vDocument.withCustomRange(null, range => {
+                            editor.vDocument.insertText('c', range);
+                            editor.renderer.render(editor.vDocument, editor.editable);
+                        });
+                    },
+                    contentAfter: 'abc[]',
+                });
+            });
+            describe('rangeParam with RangeFromTo', () => {
+                it('should work with without rangeParams.startPosition and rangeParams.endPosition', () => {
+                    testEditor({
+                        contentBefore: 'ab[]',
+                        stepFunction: async (editor: JWEditor) => {
+                            const aNode = editor.vDocument.root.next(node => node.value === 'a');
+                            const rangeParams: RangeFromTo = {
+                                start: aNode,
+                                end: aNode,
+                            };
+                            await editor.vDocument.withCustomRange(rangeParams, range => {
+                                editor.vDocument.insertText('c', range);
+                                editor.renderer.render(editor.vDocument, editor.editable);
+                            });
+                        },
+                        contentAfter: 'cab[]',
+                    });
+                });
+                it('should work with without rangeParams.end', () => {
+                    testEditor({
+                        contentBefore: 'ab[]',
+                        stepFunction: async (editor: JWEditor) => {
+                            const aNode = editor.vDocument.root.next(node => node.value === 'a');
+                            const rangeParams: RangeFromTo = {
+                                start: aNode,
+                            };
+                            await editor.vDocument.withCustomRange(rangeParams, range => {
+                                editor.vDocument.insertText('c', range);
+                                editor.renderer.render(editor.vDocument, editor.editable);
```

Comment:
Now that we have `execCommand`, I see no reason to use any other way of executing comands, especially considering `execCommand` includes the call to `render`.